### PR TITLE
DAML Engine: Use while loop instead of for loops

### PR DESCRIPTION
**Pull request by hurryabit**
This PR replaces a few `for (... to/until ...)` loops with `while`
loops which explicitly maintain their counter. In my benchmarks, this
has lead to a 3% performance improvement. Since the affected functions
are the ones used for applying arguments to functions, similar
improvements should be observable regardless of the specific benchmark.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/7126)

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 1]
```diff
@@ -320,10 +324,12 @@ object SExpr {
           machine.env.size + bounds.size - 1))
 
       // Start evaluating the let binders
-      for (i <- 1 until bounds.size) {
+      var i = 0
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:327]:
```suggestion
      var i = 1
```

**Comment by leonidr-da** [comment]
I'm surprised that this has an effect. What's your intuition about what's causing the improvement? My intuition was that this would be the other way around, because it is easier to unroll the loop.

**Comment by SamirTalwar-DA** [comment]
Scala has to construct range objects and emit boxed `RichInt` objects, rather than `Int` values, when you use a `for` comprehension over a range.

I was curious about this a while back, so I wrote equivalents, and there was almost twice as much bytecode generated for the `for` comprehension over the `while` loop, which was mostly allocations.

**Comment by leonidr-da** [comment]
@SamirTalwar-DA  Thanks for the explanation. As an experienced developer but a `Scala` novice I keep noticing these "warts" that are surprising without understanding the tradeoff. And there is no conversion back to an `Int` in this case?

**Comment by SamirTalwar-DA** [comment]
No, unfortunately not until [Valhalla](http://openjdk.java.net/projects/valhalla/) is in Java.

Java 8 gets around this in the Stream API by providing `IntStream.range`, `LongStream.range`, etc. to avoid boxing.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala (hunk 1 above)
Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala]: Ah, that's why the tests were failing. Good catch. Thank you.

**Pull request merged**